### DISCUSSION TITLE
Document Authentication

### DISCUSSION THREAD

**Discussion by afomera** [excid3/madmin#186] [Ideas]
👋

Something that would be good to document is how applications can implement a level of authentication for Madmin. Out of the box we produce an unauthenticated view, this is obviously not production ready.

Because authentication can be different from app to app, we'd prefer to not roll our own and to instead allow Developers to use their existing models instead.

What we'd like to see is:

- Examples on how to use basic auth with Madmin
- Examples of how to use it with Devise / Example of how to override the Madmin Controller to plugin any other system

Let us know if there's more information needed!

**Comment by excid3** [comment]
I'd also like to see documentation on how to add masquerade button to users show view. It's a common use case and it'd be great to have documented.

**Comment by excid3** [comment]
Route authentication with Devise and separate AdminUser:

```
rails g devise AdminUser
```

```
# config/routes.rb
authenticated :admin_user do
  namespace :madmin do
    # madmin routes
  end
end
```

**Comment by excid3** [comment]
Route authentication with User and `admin` flag:

```
rails g devise User admin:boolean
```

```
# config/routes.rb
authenticated :user, lambda { |u| u.admin? } do
  namespace :madmin do
    # madmin routes
  end
end
```

**Comment by excid3** [comment]
Started this in: https://github.com/excid3/madmin/blob/master/docs/authentication.md

**Comment by etagwerker** [comment]
What would this look like if one wanted to use [basic http authentication](https://api.rubyonrails.org/classes/ActionController/HttpAuthentication/Basic.html)? 😄 

I guess I would have to redefine this method: https://github.com/excid3/madmin/blob/f3aa501a086d629541acc941bc59274d0997eae0/test/dummy/app/controllers/madmin/application_controller.rb#L5-L10

**Comment by excid3** [comment]
I haven't tested it yet, but I left an example for http basic auth on line 20:

https://github.com/excid3/madmin/blob/f3aa501a086d629541acc941bc59274d0997eae0/test/dummy/app/controllers/madmin/application_controller.rb#L20
